### ISSUE TITLE
VolumeSnapshot support?

### ISSUE THREAD

**Issue by rbren** [rancher/local-path-provisioner#81]
Is it possible to support [VolumeSnapshots](https://kubernetes.io/docs/concepts/storage/volume-snapshots/), e.g. by making a copy of the directory?

**Comment by nvtkaszpir** [comment]
VolumeSnapshots are bit more complex than just simple file copying. 

If you just look for some backup process then local-path-provisioner is just using a directory/path in container - so you can copy fie but it is advised to stop container before file copying to avoid file corruptions, especially if the app stores extra data in memory (such as databases).
Something like this:
- stop app container which uses that pvc
- spawn new container and mount that pvc as read only, copy files elswhere, stop the container
- start app container

Second step could be also done from host side.

What do you do in the backup process depends on the use case, this can be something like rsync, copy, rclone....

**Comment by rbren** [comment]
Yeah I understand that implementing the VolumeSnapshot API would involve a lot of work.

I'm specifically interested in being able to run tests that use the VolumeSnapshot API. Ideally I'd be able to do this with a KIND cluster and local-path-storage.

**Comment by nvtkaszpir** [comment]
You could try to run different provisioner one provisoner on top of another, if supported :D

**Comment by haslersn** [comment]
Could this be supported if the backing filesystem on the host is btrfs?

**Comment by rbren** [comment]
Have learned a bit more since opening the issue.

In particular, stopping/starting the underlying pods is not part of the VolumeSnapshot API. The user is expected to handle all of that, esp if e.g. they're running a database. The VolumeSnapshot API simply tells the storage provider to make a backup of the volume in its current state.

So I think adding VolumeSnapshot support may indeed be as easy as creating a tarball or copy of the directory (and then unpacking that to create a PVC from a VS).

**Comment by haslersn** [comment]
> adding VolumeSnapshot support may indeed be as easy as creating a tarball or copy of the directory

Creating a tarball or copy is not atomic and therefore is not a "snapshot". Instead, a file system with snapshot support needs to be used.

**Comment by mlazowik** [comment]
Not stale

**Comment by mlazowik** [comment]
Not stale

**Comment by aslafy-z** [comment]
Still interested

**Comment by heruan** [comment]
This would be awesome to have for testing! For example CloudNativePG support [volume snapshot backups](https://cloudnative-pg.io/documentation/1.25/backup_volumesnapshot/#how-to-configure-volume-snapshot-backups) but those are difficult to test on a local environment during development, where I'm now using local-path-provisioner on my local cluster.

**Comment by aslafy-z** [comment]
This feature is still desired!

**Comment by haslersn** [comment]
This is the most requested feature of this project – please exempt it from the stale bot.

**Comment by kastl-ars** [comment]
Not stale.

Can someone please disable the stale bot?

**Comment by jimmyjones2** [comment]
This feature request is not stale

**Comment by saolof** [comment]
On recent kernels, `cp -a --reflink=always`  actually works between bind mounts to the same underlying xfs or btrfs filesystem thanks to https://www.spinics.net/lists/linux-btrfs/msg121228.html . And GNU cp will reflink by default. So this should be easily doable by just shelling out to coreutils and the hard part (where doing it here adds value) is in the control plane bookkeeping

It's atomic per file but not atomic per volume. That's still arguably better than most volumesnapshot implementations from implementations like ceph-csi or commercial SAN vendor implementations like the HPE one, which do a block level snapshot without doing an fsfreeze like the LVM cli would, and it's perfectly usable for cloudnativepg cold volumesnapshot backups for example which are very useful for development.

**Comment by aslafy-z** [comment]
This feature is still desired.
